task-maker-async/src/store.rs

```diff
@@ -0,0 +1,114 @@
+#![allow(dead_code)]
+
+use serde::{Deserialize, Serialize};
+
+type HashData = Vec<u8>;
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct InputFileHash(HashData);
+
+/// Two-level hash; the outer hash has information about all properties of a computation that are
+/// expected to change the result (such as outer hashes of inputs, or the command line). The inner
+/// hash takes care of properties of the computation that should not change the outputs, such as
+/// time and memory limits.
+#[derive(Debug, Serialize, Deserialize)]
+pub struct ComputationHash(HashData, HashData);
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct InputFileWriteHandle(usize);
+#[derive(Debug, Serialize, Deserialize)]
+pub struct ComputationWriteHandle(usize);
+#[derive(Debug, Serialize, Deserialize)]
+pub struct ReadHandle(usize);
+
+#[derive(Debug, Serialize, Deserialize)]
+pub enum ComputationDataFile {
+    Outcome,
+    OutputFile(String),
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub enum FileReadingOutcome {
+    /// The file has been deleted, for example because the worker responsible for the execution has
+    /// disappeared.
+    Dropped,
+    /// The file has been fully read.
+    EndOfFile,
+    /// A new chunk of data is available.
+    Data(Vec<u8>),
+}
+
+#[tarpc::service]
+pub trait Store {
+    /// Used by the client to upload inputs to the DAG. Returns None if the file is already
+    /// present. Either way, this leases the corresponding hash for LEASE_LENGTH_SEC.
+    async fn create_input_file(hash: InputFileHash) -> Option<InputFileWriteHandle>;
+
+    /// Appends a chunk to the input file; the InputFileWriteHandle must have been obtained through
+    /// create_input_file. Renews the lease on the corresponding hash.
+    async fn append_input_chunk(file_handle: InputFileWriteHandle, data: Vec<u8>)
+        -> Result<(), ()>;
+
+    /// Finalizes the input file; no more data can be appended, and the resulting hash must match
+    /// the claimed hash when create_input_file was called.
+    async fn finalize_input_write_file(file_handle: InputFileWriteHandle) -> Result<(), ()>;
+
+    /// Refreshes the lease for the given input file, or creates one if none was present.
```

Comment:
I think this is not the same kind of lease that append_input_chunk is talking about. If so, please name them in a distinguishuable way.

Reply:
The (a) lack of distinction between creation and refresh and (b) lack of lease identity make it easier for bugs where leases aren't refreshed to hide if someone else is usually keeping a lease on the same file at the same time (or just if the GC doesn't manage to get rid of the file in time).

Maybe have a handle for a lease and refresh/create one separately?

Reply:
Right. Modified to only refresh an input write lease, identified by a file_handle. (read_input_file gives you the option to get a read-only lease).